File: debop4s-data/src/test/scala/com/github/debop4s/data/tests/mapping/embeddable/EmbeddableTest.scala
package com.github.debop4s.data.tests.mapping.embeddable

import com.github.debop4s.core.ValueObject
import com.github.debop4s.core.utils.Hashs
import com.github.debop4s.data.model.LongEntity
import com.github.debop4s.data.tests.AbstractJpaTest
import javax.persistence._
import org.hibernate.annotations.{DynamicUpdate, DynamicInsert}
import org.junit.Test

/**
 * Created by debop on 2014. 3. 9.
 */
@org.springframework.transaction.annotation.Transactional
class EmbeddableTest extends AbstractJpaTest {

  @PersistenceContext val em: EntityManager = null

  @Test
  def embeddableAddress() {
    val user = new EmbeddableUser()
    user.username = "debop"
    user.password = "1234"
    user.email = "debop@gmail.com"

    user.homeAddress.city = "서울"
    user.homeAddress.street = "정릉"
    user.homeAddress.zipcode = "100-200"

    user.officeAddress.city = "서울"
    user.officeAddress.street = "안국"
    user.officeAddress.zipcode = "100-999"

    em.persist(user)
    em.flush()
    em.clear()

    val loaded = em.find(classOf[EmbeddableUser], user.id)
    assert(loaded != null)
    assert(loaded.homeAddress == user.homeAddress)
    assert(loaded.officeAddress == user.officeAddress)

    em.remove(loaded)
    em.flush()

    assert(em.find(classOf[EmbeddableUser], user.id) == null)
  }
}

@Embeddable
@DynamicInsert
@DynamicUpdate
class EmbeddableAddress extends ValueObject {

  var street: String = _
  var zipcode: String = _
  var city: String = _

  override def hashCode(): Int = Hashs.compute(street, zipcode, city)
}

@Entity
@Index(name = "idx_user_username", columnList = "username, password")
class EmbeddableUser extends LongEntity {

  var firstname: String = _
  var lastname: String = _

  @Column(length = 128, nullable = false)
  var username: String = _

  @Column(length = 64, nullable = false)
  var password: String = _

  @Column(length = 128, nullable = false)
  @Index(name = "ix_user_email", columnList = "email")
  var email: String = _

  var active: Boolean = true

  @Embedded
  @AttributeOverrides(Array(
    new AttributeOverride(name = "street", column = new Column(name = "HomeStreet", length = 128)),
    new AttributeOverride(name = "zipcode", column = new Column(name = "HomeZipCode", length = 24)),
    new AttributeOverride(name = "city", column = new Column(name = "HomeCity", length = 128))
  ))
  var homeAddress = new EmbeddableAddress()

  @Embedded
  @AttributeOverrides(Array(
    new AttributeOverride(name = "street", column = new Column(name = "OfficeStreet", length = 128)),
    new AttributeOverride(name = "zipcode", column = new Column(name = "OfficeZipCode", length = 24)),
    new AttributeOverride(name = "city", column = new Column(name = "OfficeCity", length = 128))
  ))
  var officeAddress = new EmbeddableAddress()

  override def hashCode(): Int = Hashs.compute(username, password)
}

File: debop4s-data-orm/src/test/java/debop4s/data/orm/jtests/mapping/embeddable/EmbeddableTest.java
package debop4s.data.orm.jtests.mapping.embeddable;

import debop4s.data.orm.jtests.jpa.JpaTestBase;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.data.orm.s.mapping.embeddable.EmbeddableJUnitSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 7. 22. 오후 5:23
 */
@Slf4j
@Transactional
public class EmbeddableTest extends JpaTestBase {

    @PersistenceContext
    EntityManager em;

    @Test
    public void embeddableAddressTest() {
        EmbeddableUser user = new EmbeddableUser();

        user.setUsername("배성혁");

        user.getHomeAddress().setCity("서울");
        user.getHomeAddress().setStreet("정릉로");
        user.getHomeAddress().setZipcode("100-100");

        user.getOfficeAddress().setCity("서울");
        user.getOfficeAddress().setStreet("안국로");
        user.getOfficeAddress().setZipcode("200-200");

        em.persist(user);
        em.flush();
        em.clear();

        EmbeddableUser loaded = em.find(EmbeddableUser.class, user.getId());

        assertThat(loaded).isNotNull();
        assertThat(loaded.getHomeAddress().getZipcode()).isEqualTo(user.getHomeAddress().getZipcode());
        assertThat(loaded.getOfficeAddress().getZipcode()).isEqualTo(user.getOfficeAddress().getZipcode());

        em.remove(loaded);
        em.flush();

        assertThat(em.find(EmbeddableUser.class, user.getId())).isNull();
    }
}
